### PR TITLE
Implement a bunch of pickle serialization features that optimize for size.

### PR DESCRIPTION
**Pull request by resistor**
This saves about 10KB of compressed size on FaceBlaze.  https://github.com/pytorch/pytorch/issues/23582

### CODE REVIEW

**Comment by resistor** [comment]
@driazati looking for pointers on testing for this.

File: torch/csrc/jit/pickler.cpp [hunk 1]
```diff
@@ -202,7 +202,8 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
 
 void Pickler::pushIValue(const IValue& ivalue) {
   // Check if reference ivalue has been saved before
-  if (ivalue.isPtrType()) {
+  // NOTE: Strings are always memoized by value rather than by pointer.
```
Comment (jamesr66a) [inline torch/csrc/jit/pickler.cpp]: s/Strings/strings and tuples?

File: torch/csrc/jit/pickler.cpp [hunk 2]
```diff
@@ -218,18 +219,23 @@ void Pickler::pushIValue(const IValue& ivalue) {
...
     memoized_ivalue_map_[ivalue.internalToPointer()] = pushNextBinPut();
   }
 }
 
 void Pickler::pushInt(const IValue& ivalue) {
   auto n = ivalue.toInt();
-  if (n >= std::numeric_limits<int8_t>::min() &&
-      n <= std::numeric_limits<int8_t>::max()) {
+  if (n >= std::numeric_limits<uint8_t>::min() &&
+      n <= std::numeric_limits<uint8_t>::max()) {
```
Comment (jamesr66a) [inline torch/csrc/jit/pickler.cpp]: Why this change?

Comment (resistor) [reply torch/csrc/jit/pickler.cpp]: Because the old version was incorrect.  `BININT1` can only encode positive values: https://github.com/python/cpython/blob/master/Lib/pickletools.py#L1196

**Review by driazati — COMMENTED** [review]
Where are the serialization savings coming from here? Were there a bunch of unnecessary `BINPUT` codes? The tuple specializations only save a byte each for the `MARK` as well, so I'm curious how that adds up to 10 KB.

As for testing, you could save an archive to a buffer and check the size of the file (see below for a similar example). We shouldn't be regressing in size for a known set of attributes so a magic number (with an explanation) for the expected size is probably fine.

https://github.com/pytorch/pytorch/blob/489cc46686e7e689e7130e78b73fe77a9edac5ed/test/test_jit.py#L3474

**Comment by resistor** [comment]
> Where are the serialization savings coming from here? Were there a bunch of unnecessary `BINPUT` codes? The tuple specializations only save a byte each for the `MARK` as well, so I'm curious how that adds up to 10 KB.

The biggest benefit comes from memoizing strings by value rather than by pointer.  The next biggest benefit comes from not memoizing tuples by pointer, eliminating a bunch of ununsed `BINPUT` codes.  (There is more potential gain is we memoize tuples by value, but I haven't figured out a good implementation yet).  Implementing the various `TUPLE` codes is the smallest benefit, but was still worth ~1KB as there are a *lot* of tuples going around.

**Comment by resistor** [comment]
> As for testing, you could save an archive to a buffer and check the size of the file (see below for a similar example). We shouldn't be regressing in size for a known set of attributes so a magic number (with an explanation) for the expected size is probably fine.

This doesn't work, because ZIP compression is applied after pickling and it becomes impossible to directly test the behavior of the pickler vs the behavior of the DEFLATE compression.

**Comment by driazati** [comment]
> This doesn't work, because ZIP compression is applied after pickling and it becomes impossible to directly test the behavior of the pickler vs the behavior of the DEFLATE compression.

You could do a C++ test and use #23241 once it lands

File: torch/csrc/jit/pickler.cpp [hunk 3]
```diff
@@ -218,18 +219,23 @@ void Pickler::pushIValue(const IValue& ivalue) {
     }
   }
   pushIValueImpl(ivalue);
-  if (ivalue.isPtrType()) {
+  if (ivalue.isPtrType() && !ivalue.isString() && !ivalue.isTuple()) {
```
Comment (Krovatkin) [inline torch/csrc/jit/pickler.cpp]: this exact check is also repeated below. Maybe, we should consider moving it into a helper  (`isEfficientlyMemoizablePtrType`) or moving the whole `isPtrType` case into its own function and just calling `pushIValueImpl`, otherwise. Even though it looks like the second check is what that gives us size savings, refactoring it into the helper might help us avoid miscopying it in the future

File: torch/csrc/jit/pickler.cpp [hunk 4]
```diff
@@ -218,18 +219,23 @@ void Pickler::pushIValue(const IValue& ivalue) {
...
   auto n = ivalue.toInt();
-  if (n >= std::numeric_limits<int8_t>::min() &&
-      n <= std::numeric_limits<int8_t>::max()) {
+  if (n >= std::numeric_limits<uint8_t>::min() &&
+      n <= std::numeric_limits<uint8_t>::max()) {
     push<OpCode>(OpCode::BININT1);
-    push<int8_t>(n);
+    push<uint8_t>(n);
+  } else if (
+      n >= std::numeric_limits<uint16_t>::min() &&
```
Comment (Krovatkin) [inline torch/csrc/jit/pickler.cpp]: nice catch!

File: torch/csrc/jit/pickler.cpp [hunk 5]
```diff
@@ -652,6 +683,28 @@ OpCode Unpickler::readInstruction() {
...
     } break;
+    case OpCode::TUPLE1: {
+        auto tuple = c10::ivalue::Tuple::create({stack_.back()});
+        stack_.pop_back();
+        stack_.emplace_back(tuple);
+    } break;
+    case OpCode::TUPLE2: {
+        auto stack_size = stack_.size();
+        auto tuple = c10::ivalue::Tuple::create(
+          {stack_[stack_size-2], stack_[stack_size-1]});
```
Comment (Krovatkin) [inline torch/csrc/jit/pickler.cpp]: maybe, we could use `pop(stack, 2)` here?

File: torch/csrc/jit/pickler.cpp [hunk 6]
```diff
@@ -652,6 +683,28 @@ OpCode Unpickler::readInstruction() {
...
+        stack_.pop_back();
+        stack_.pop_back();
+        stack_.emplace_back(tuple);
+    } break;
+    case OpCode::TUPLE3: {
+        auto stack_size = stack_.size();
+        auto tuple = c10::ivalue::Tuple::create(
+          {stack_[stack_size-3], stack_[stack_size-2], stack_[stack_size-1]});
+        stack_.pop_back();
```
Comment (Krovatkin) [inline torch/csrc/jit/pickler.cpp]: ditto

**Review by Krovatkin — COMMENTED** [review]
added a few comments.

File: torch/csrc/jit/pickler.cpp [hunk 7]
```diff
@@ -201,8 +201,12 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
 }
 
 void Pickler::pushIValue(const IValue& ivalue) {
-  // Check if reference ivalue has been saved before
-  if (ivalue.isPtrType()) {
+  bool isMutableType = ivalue.isPtrType() && !ivalue.isString() && !ivalue.isTuple();
```
Comment (zdevito) [inline torch/csrc/jit/pickler.cpp]: Memoization of tuples will change the behavior of `is` on Tuples. I am not sure whether this is an allowed semantic change. It is probably safer not to do it.

```
import pickle

a = (3, 4)
b = (a, a)

saved = pickle.dumps(b)
b_ = pickle.loads(saved)

print(b_[0] is b_[1]) # True
```

**Review by zdevito — COMMENTED** [review]
Changes to int/tuple encoding look good, but I am skeptical about not memoizing tuples. First it changes the behavior of `is`, second, it has bad worst-case behavior. If N of really small objects share a reference to a really huge tuple of size M, we are going to store N*M values when we should be storing N + M.

**Comment by zdevito** [comment]
There might be some other way to remove BINPUT codes. If the tuples that generate them have refcount one, we can safely omit the BINPUT since we no there is not a possibility of a BINGET.

**Review by zdevito — APPROVED** [review]
